### DISCUSSION TITLE
Request failed with status code 403: Unauthorized.

### DISCUSSION THREAD

**Discussion by haris-pixleedge** [opcodesio/log-viewer#353] [Q&A]
Log viewer working fine on localhost but get Unauthorized error on production although route is working but can't see the log file due to error. 

I've also added the Log viewer Auth to restrict the access in AppServiceProvider:

` LogViewer::auth(function ($request) {
            return $request->user()
                && in_array($request->user()->email, [
                    'haris.zahid@example.com',
                ]);
        });`

![image](https://github.com/opcodesio/log-viewer/assets/105644471/522b4c0d-65c4-45b4-9a63-d8ad082095bb)

Here is my config/log-viewer file:

`<?php

return [

    /*
    |--------------------------------------------------------------------------
    | Log Viewer
    |--------------------------------------------------------------------------
    | Log Viewer can be disabled, so it's no longer accessible via browser.
    |
    */

    'enabled' => env('LOG_VIEWER_ENABLED', true),

    'require_auth_in_production' => true,

    /*
    |--------------------------------------------------------------------------
    | Log Viewer Domain
    |--------------------------------------------------------------------------
    | You may change the domain where Log Viewer should be active.
    | If the domain is empty, all domains will be valid.
    |
    */

    'route_domain' => null,

    /*
    |--------------------------------------------------------------------------
    | Log Viewer Route
    |--------------------------------------------------------------------------
    | Log Viewer will be available under this URL.
    |
    */

    'route_path' => 'log-viewer',

    /*
    |--------------------------------------------------------------------------
    | Back to system URL
    |--------------------------------------------------------------------------
    | When set, displays a link to easily get back to this URL.
    | Set to `null` to hide this link.
    |
    | Optional label to display for the above URL.
    |
    */

    'back_to_system_url' => config('app.url', null),

    'back_to_system_label' => null, // Displayed by default: "Back to {{ app.name }}"

    /*
    |--------------------------------------------------------------------------
    | Log Viewer time zone.
    |--------------------------------------------------------------------------
    | The time zone in which to display the times in the UI. Defaults to
    | the application's timezone defined in config/app.php.
    |
    */

    'timezone' => null,

    /*
    |--------------------------------------------------------------------------
    | Log Viewer route middleware.
    |--------------------------------------------------------------------------
    | Optional middleware to use when loading the initial Log Viewer page.
    |
    */

    'middleware' => [
        'web',
        \Opcodes\LogViewer\Http\Middleware\AuthorizeLogViewer::class,
    ],

    /*
    |--------------------------------------------------------------------------
    | Log Viewer API middleware.
    |--------------------------------------------------------------------------
    | Optional middleware to use on every API request. The same API is also
    | used from within the Log Viewer user interface.
    |
    */

    'api_middleware' => [
        \Opcodes\LogViewer\Http\Middleware\EnsureFrontendRequestsAreStateful::class,
        \Opcodes\LogViewer\Http\Middleware\AuthorizeLogViewer::class,
    ],

    /*
    |--------------------------------------------------------------------------
    | Log Viewer Remote hosts.
    |--------------------------------------------------------------------------
    | Log Viewer supports viewing Laravel logs from remote hosts. They must
    | be running Log Viewer as well. Below you can define the hosts you
    | would like to show in this Log Viewer instance.
    |
    */

    'hosts' => [
        'local' => [
            'name' => ucfirst(env('APP_ENV', 'local')),
        ],

        // 'staging' => [
        //     'name' => 'Staging',
        //     'host' => 'https://staging.example.com/log-viewer',
        //     'auth' => [      // Example of HTTP Basic auth
        //         'username' => 'username',
        //         'password' => 'password',
        //     ],
        // ],
        //
        // 'production' => [
        //     'name' => 'Production',
        //     'host' => 'https://example.com/log-viewer',
        //     'auth' => [      // Example of Bearer token auth
        //         'token' => env('LOG_VIEWER_PRODUCTION_TOKEN'),
        //     ],
        //     'headers' => [
        //         'X-Foo' => 'Bar',
        //     ],
        // ],
    ],

    /*
    |--------------------------------------------------------------------------
    | Include file patterns
    |--------------------------------------------------------------------------
    |
    */

    'include_files' => [
        '*.log',
        '**/*.log',

        // You can include paths to other log types as well, such as apache, nginx, and more.
        '/var/log/httpd/*',
        '/var/log/nginx/*',

        // MacOS Apple Silicon logs
        '/opt/homebrew/var/log/nginx/*',
        '/opt/homebrew/var/log/httpd/*',
        '/opt/homebrew/var/log/php-fpm.log',
        '/opt/homebrew/var/log/postgres*log',
        '/opt/homebrew/var/log/redis*log',
        '/opt/homebrew/var/log/supervisor*log',

        // '/absolute/paths/supported',
    ],

    /*
    |--------------------------------------------------------------------------
    | Exclude file patterns.
    |--------------------------------------------------------------------------
    | This will take precedence over included files.
    |
    */

    'exclude_files' => [
        // 'my_secret.log'
    ],

    /*
    |--------------------------------------------------------------------------
    | Hide unknown files.
    |--------------------------------------------------------------------------
    | The include/exclude options above might catch files which are not
    | logs supported by Log Viewer. In that case, you can hide them
    | from the UI and API calls by setting this to true.
    |
    */

    'hide_unknown_files' => true,

    /*
    |--------------------------------------------------------------------------
    |  Shorter stack trace filters.
    |--------------------------------------------------------------------------
    | Lines containing any of these strings will be excluded from the full log.
    | This setting is only active when the function is enabled via the user interface.
    |
    */

    'shorter_stack_trace_excludes' => [
        '/vendor/symfony/',
        '/vendor/laravel/framework/',
        '/vendor/barryvdh/laravel-debugbar/',
    ],

    /*
    |--------------------------------------------------------------------------
    | Cache driver
    |--------------------------------------------------------------------------
    | Cache driver to use for storing the log indices. Indices are used to speed up
    | log navigation. Defaults to your application's default cache driver.
    |
    */

    'cache_driver' => env('LOG_VIEWER_CACHE_DRIVER', null),

    /*
    |--------------------------------------------------------------------------
    | Chunk size when scanning log files lazily
    |--------------------------------------------------------------------------
    | The size in MB of files to scan before updating the progress bar when searching across all files.
    |
    */

    'lazy_scan_chunk_size_in_mb' => 50,

    'strip_extracted_context' => true,
];
`

**Comment by haris-pixleedge** [comment]
![image](https://github.com/opcodesio/log-viewer/assets/105644471/82208ef6-c403-4cbc-80a9-0afc56a386ca)

**Comment by arukompas** [comment]
Hey @haris-pixleedge 

From my experience this is most likely to do with "stateful domains". By default, it will only allow access from the domain (or subdomain) defined in the `APP_URL` environment variable. Keep in mind that `www.example.com` and `example.com` are considered different.

Here's what you should do:

1. Update Log Viewer to **v3.7.0** (or later)
2. Add a new environment variable which contains comma-separated list of domains or subdomains where your app is accessible from, e.g.:
```env
LOG_VIEWER_API_STATEFUL_DOMAINS=example.com,www.example.com
```

OR, just update your `APP_URL` to the correct domain that you're accessing the app from.

Let me know how it goes!

**Reply by artem-dev2** [reply]
Tried adding `LOG_VIEWER_API_STATEFUL_DOMAINS` with current domain does not help, in the APP_URL also set as current domain...

`log-viewer/api/folders?direction=desc` this call doesn not go well it's erroring with 400 status code.

I'm on 3.8 version

**Reply by arukompas** [reply]
hey @artem-dev2 , could you please share the network tab screenshot of the failing request? Please make sure the request and response headers are visible.

Also, just double-checking - what's the URL to the log viewer in the browser, and what is the value of `LOG_VIEWER_API_STATEFUL_DOMAINS` ?

**Reply by artem-dev2** [reply]
I sorted that out, the problem came from middlawares those two in config:

```
        //Opcodes\LogViewer\Http\Middleware\EnsureFrontendRequestsAreStateful::class,
        //Opcodes\LogViewer\Http\Middleware\AuthorizeLogViewer::class,
```

not sure for what are they, but just commended out and now it works fine.

**Reply by arukompas** [reply]
@artem-dev2 Right, but now your Log Viewer is probably not protected and accessible by anyone.

Make sure you define an auth callback in production. Locally it should be fine, of course.

**Reply by haris-pixleedge** [reply]
@arukompas How to update Log Viewer to v3.7.0?

**Reply by arukompas** [reply]
@haris-pixleedge just this command:

```shell
composer update opcodes/log-viewer
```

**Reply by mhd-fettah** [reply]
updating "APP_URL" was the issue . for me . 
thanks lots.

**Reply by dequilla3** [reply]
> updating "APP_URL" was the issue . for me . thanks lots.

this solved my problem

**Comment by jjpergui** [comment]
Hey @arukompas 

I had a similar problem with Laravel 10 / PHP 8.2. I kept getting error 500 while loading the index page:

![Captura de Pantalla 2024-09-06 a las 20 18 40](https://github.com/user-attachments/assets/857cd4a8-3b5d-4e14-b0dd-64341dc4fe5a)

I had to disable **\Opcodes\LogViewer\Http\Middleware\EnsureFrontendRequestsAreStateful::class,** in the log-viewer.php file to get it to work.

These are the traces of the 500 error:

![Captura de Pantalla 2024-09-06 a las 20 19 15](https://github.com/user-attachments/assets/de69ee95-8012-44d0-b0e9-9e99b892ac45)

![Captura de Pantalla 2024-09-06 a las 20 19 40](https://github.com/user-attachments/assets/cd1fa530-2af7-45ae-9a62-bdd29cecbd23)

After deactivating that middleware, everything worked fine.

Thanks for your great job. Next step is makint it multi-language!

Jordi

**Comment by dhaliscak** [comment]
I had a same problem, I was able to open /log-viewer, but giving me 403. I fixed it **partially**.


I have updated my log-viewer package, but my `config/log-viewer.php` contained just a few atributes: `route_path, middleware, include_files`. Nothing else. So I copied config from `vendor/opcodesio/log-viewer/config/`, and updated atributes above to match my website.


I had to remove **AuthorizeLogViewer**, if it is there, I am still getting 403. So I replaced it with **auth**, **role:Admin**. Now I am able to access the page, but in that logs list on left it is still 403. So it is failing now on API.
```
'middleware' => [
            'web',
            'auth',
            'role:Admin',
            // \Opcodes\LogViewer\Http\Middleware\AuthorizeLogViewer::class,
    ],
```



If I comment out **AuthorizeLogViewer** also there, **it is working fine**, but if I open API url in incognito, I can see JSON data:
```
'api_middleware' => [
        \Opcodes\LogViewer\Http\Middleware\EnsureFrontendRequestsAreStateful::class,
        // \Opcodes\LogViewer\Http\Middleware\AuthorizeLogViewer::class,
    ],
```

My .env:
```
APP_URL=https://example.com
LOG_VIEWER_API_STATEFUL_DOMAINS=example.com,www.example.com
```

Logs are accesible from https://example.com/log-viewer, nothing crazy, no subdomains.

I understand that logs are not protected now, but I cannot figure out how to make it work. 😵

_Installed Version: v3.12.0_

**Comment by martinpl** [comment]
For me issue was that ```$request->user()``` wasn't available in API route,
Adding those middlewares to api_middleware fixed that: 
```
\Illuminate\Cookie\Middleware\EncryptCookies::class,
\Illuminate\Session\Middleware\StartSession::class,
```

**Comment by alaspinaIDL** [comment]
I use this middleware from Opcodes\LogViewer\Http\Middleware\ForwardRequestToHostMiddleware

```php
<?php

namespace App\Http\Middleware;

use Closure;
use Illuminate\Support\Str;
use Illuminate\Support\Facades\Http;
use Opcodes\LogViewer\Facades\LogViewer;

class LogViewerApi
{
    /**
     * Handle an incoming request.
     *
     * @param  \Illuminate\Http\Request  $request
     * @param  \Closure  $next
     * @param  string|null  $guard
     * @return mixed
     */
    public function handle($request, Closure $next, $guard = null)
    {
        $query = $request->query();
        $hostIdentifier = $query['host'] ?? '';
        unset($query['host']);
        $host = LogViewer::getHost($hostIdentifier);

        if ($host) {
            $actionPath = Str::replaceFirst(config('log-viewer.route_path'), '', $request->path());
            $url = $host->host . $actionPath . (! empty($query) ? '?' . http_build_query($query) : '');
            $headers = array_merge([
                'X-Forwarded-Host' => $request->getHost(),
                'X-Forwarded-Port' => $request->getPort(),
                'X-Forwarded-Proto' => $request->getScheme(),
            ], $host->headers ?? []);

            $proxyRequest = Http::withHeaders($headers)->acceptJson();

            if ($host->auth && isset($host->auth['token'])) {
                $proxyRequest = $proxyRequest->withToken($host->auth['token']);
            } elseif ($host->auth && isset($host->auth['username']) && isset($host->auth['password'])) {
                $proxyRequest = $proxyRequest->withBasicAuth($host->auth['username'], $host->auth['password']);
            } elseif ($host->auth && isset($host->auth['digest'])) {
                $proxyRequest = $proxyRequest->withDigestAuth($host->auth['username'], $host->auth['password']);
            }

            if (! $host->verifyServerCertificate) {
                $proxyRequest = $proxyRequest->withoutVerifying();
            }

            $response = $proxyRequest->send($request->method(), $url);

            return response(
                $response->body(),
                $response->status(),
                [
                    'Content-Type' => $response->header('Content-Type'),
                ]
            );
        }

        return $next($request);
    }
}
```
and in log-viewer.php <----- config

```php
'hosts' => [
        'local' => [
            'name' => 'Local',
            'host' => env('APP_URL', 'http://localhost') . '/log-viewer',
        ],
        'production' => [
            'name' => 'Production',
            'host' => env('APP_URL', 'https://example.domain.com') . '/log-viewer',
            'headers' => [
                'apikey' => env('LOG_VIEWER_APIKEY'),
            ],
            'verify_server_certificate' => false,
        ],
    ],
```
    and i have other middleware working with 'apikey' header for multiples purposes

**Comment by chirag-vaghamashi** [comment]
We are having the same issue with the latest version for the production environment.

It's fixed by adding a domain array to the variable below in the config file.

`LOG_VIEWER_API_STATEFUL_DOMAINS=['test.com']`

**Reply by mankowitz** [reply]
I don't think you need the brackets. You should be able to do 

`LOG_VIEWER_API_STATEFUL_DOMAINS=test1.com,test2.com`

**Comment by icanall10** [comment]
nothing worked for me.
but changing octane roadrunner instead of frankenphp worked
